### ISSUE TITLE
Logstash refuses to start if no database-connection can be made

### ISSUE THREAD

**Issue by scaarup** [logstash-plugins/logstash-input-jdbc#76]
The error reported is:
  Java::JavaSql::SQLException: JZ006: Caught IOException: java.net.ConnectException: Connection refused

In my opinion this should not be a complete show-stopper for logstash, as I have several other working inputs. A jdbc connection should be attempted as per defined schedule.

**Comment by purbon** [comment]
@scaarup thanks a lot for your feedback, I'm only wondering one thing to understand your request. So your expectation would be to try to open the connection every time a query is going to be made? I understand keeping it open and reconnecting in case it goes down. Is this summarizing your issue?

Looking forward to know more,

**Comment by phkr** [comment]
I just ran in the same situation as @scaarup. In my case, the target database was available when I configured logstash. The database was shut down while logstash was running. When I tried to restart logstash, it did not come up again because it could not connect to the database.

I have the same opinion as @scaarup - my other inputs are fine and the database will eventually be up and running again. It would be good if logstash could just start up and the jdbc connections would be reattempted in regular intervals (for example as per defined schedule) until the database is back up again.

**Comment by jbehl** [comment]
@purbon I also have encountered this and believe it is a bug as well.  As is the same with @scaarup, I cannot proceed with using the plugin unfortunately due to this.  

A connection should be re-attempted every scheduled interval.  Just because it is not available at the time should not cause logstash to exit.  For example, if a "file" input is created and the file does not exist, this does not connotate an error condition.  The same should be true of the database being available for query or not.

**Comment by veshtov** [comment]
I agree with the above. This should be addressed. In the case of several DB pulls and having potential alternate DB's for the same servers (due to configuration constraints of the application teams) it is not always possible to have a VIP or something placed in front, especially with legacy systems. 

It would be great if this would allow us to have all DB's setup without LS failing upon connection.

**Comment by BellBiker** [comment]
I recognized the same issue that logstash crashes if the database connection is not available at starttime. But it also causes trouble in logstash 2.2. and also 2.3. if a database is shutdown or unreachable (eg. to generate a backup). During the time of unavailability I got the following error message:
`{:timestamp=>"2016-06-01T05:40:00.255000+0200", :message=>"Java::JavaSql::SQLRecoverableException: Keine weiteren Daten aus Socket zu lesen: select ..... order by 2 desc", :level=>:error}`
If the database is restarted I got a different error message 
`{:timestamp=>"2016-06-01T05:42:00.142000+0200", :message=>"Java::JavaSql::SQLRecoverableException: I/O-Fehler: Software caused connection abort: socket write error: select .... order by 2 desc", :level=>:error}`
and afterwards logstash does not try to read data from the database as scheduled. Only after restarting logstash (hopefully without any other unavailable database) it runs as scheduled.

**Comment by scaarup** [comment]
@purbon yes "keeping it open and reconnecting in case it goes down" is the summary of my issue. Thank you.

**Comment by purbon** [comment]
I tested the reconnect locally today with mysql and raised the error, but then when the database was back everything was good and in order, I'm guessing this might be dependant of your jdbc driver?

**Comment by veshtov** [comment]
This happens with standard jdbc driver as well as custom ones. The issue is that Logstash can crash and fail to continue with rufus scheduled tasks if JDBC connection is not online. The service must be restarted. 

reconnect will attempt for the prescribed amount of time. The issue is that we would like this to be more resilient and to not lock on error. It should continue and re-attempt on the next rufus schedule point.

**Comment by PhaedrusTheGreek** [comment]
Also see https://github.com/logstash-plugins/logstash-input-jdbc/issues/144

I could **not** reproduce the issue of crashing once Logstash has started, but can easily reproduce logstash failing to start on JDBC connection issues.

**Comment by veshtov** [comment]
Thanks Phaedrus, 

I'll see if I can pinpoint more info in logs if it happens again. We mostly see the crashing once it is started with Oracle based systems. So if the datagaurd instance (Oracle) is failed over to another active server the connection fails and will cause logstash to crash. 

we get the fail to start issue on mssql, oracle, or postgres when the DB is down.

**Comment by minibbjd** [comment]
I run into the same problem with logstash 5.1 and jdbc inputs from Oracle databases. 

This is a showstopper problem for using jdbc inputs with logstash.
